src/legacy/core_plugins/kibana/public/dashboard/np_ready/application.ts

```diff
@@ -67,21 +73,57 @@ export interface RenderDeps {
   embeddables: IEmbeddableStart;
   localStorage: Storage;
   share: SharePluginStart;
+
+  // hack:
+  // renderApp() called each time the app is mounted,
+  // but initialisation of angular module happens only once.
+  // On each app mount, new history and kbnUrlStateStorage instances are created
+  // and we have to make sure, that those new instances will be used in DashboardAppController
+  // and not the ones which where created when angular module was initialised.
+  // This is achieved by having reference to initial object, which angular module was initialised with,
+  // and then by mutating that object on subsequent mounts we can pass new instances down to controller.
+  // If not for this hack, we could end up that global state syncing uses different history instance then app state syncing
+  // Which could cause excessive browser history entries.
+  angularGlobalStateHacks?: AngularGlobalStateHacks;
 }
 
+interface AngularGlobalStateHacks {
+  hasInheritedGlobalState?: boolean;
+  kbnUrlStateStorage?: IKbnUrlStateStorage;
+  history?: History;
+}
 let angularModuleInstance: IModule | null = null;
+const angularGlobalStateHacks: AngularGlobalStateHacks = {};
 
 export const renderApp = (element: HTMLElement, appBasePath: string, deps: RenderDeps) => {
+  const history = createHashHistory();
```

Comment:
nit: We could put that part into a function `initGlobalStateHandling` that is called in the dashboard controller in `src/legacy/core_plugins/kibana/public/dashboard/np_ready/dashboard_app_controller.tsx` and in the listing controller in `src/legacy/core_plugins/kibana/public/dashboard/np_ready/legacy_app.js` - `initGlobalStateHandling` would return a function that stops the syncing and can be registered within the controller using `    $scope.$on('$destroy', () => {`. Then we don't need a reference on the application level and don't need to overwrite typescripts type checks

Reply:
thanks, looks better 👍 
Hope I got it right.